---
- name: Create /opt/haproxy directory
  file:
    path: /opt/haproxy
    state: directory
  tags: reconfigure

- name: Configure External SSL
  block:
    - name: Configure Let's Encrypt certificates
      import_tasks: ssl/letsencrypt.yml
      when:
        - external_url_ssl
        - external_ssl_source == 'letsencrypt'

    - name: Configure User provided certificates
      import_tasks: ssl/user.yml
      when:
        - external_url_ssl
        - external_ssl_source == 'user'

    - name: Cleanup any previous External SSL config if disabled
      import_tasks: ssl/none.yml
      when:
        - not external_url_ssl
        - "external_ssl_source == ''"
  when: "'haproxy_external' in group_names"
  tags: reconfigure

- name: Create /opt/haproxy/haproxy.cfg file
  template:
    src: templates/haproxy.cfg.j2
    dest: /opt/haproxy/haproxy.cfg
  tags: reconfigure

- name: Start HAProxy Internal Docker
  docker_container:
    name: haproxy
    image: haproxy:alpine
    pull: true
    recreate: true
    restart_policy: unless-stopped
    state: started
    ports: ['80:80', '6432:6432', '2305:2305', '1936:1936', '5432:5432']
    volumes:
      - /opt/haproxy:/usr/local/etc/haproxy:ro
  when: "'haproxy_internal' in group_names"
  tags:
    - reconfigure
    - restart

- name: Start HAProxy External Docker
  docker_container:
    name: haproxy
    image: haproxy:alpine
    pull: true
    recreate: true
    restart_policy: unless-stopped
    state: started
    ports: ['80:80', '443:443', '1936:1936', '2222:2222', '9090:9090', '9122:9122', '5601:5601']
    volumes:
      - /opt/haproxy:/usr/local/etc/haproxy:ro
  when: "'haproxy_external' in group_names"
  tags:
    - reconfigure
    - restart

- name: Wait for HAProxy Docker to be running
  shell: docker inspect -f \{\{.State.Status\}\} haproxy
  register: haproxy_docker_status
  until: haproxy_docker_status.stdout == 'running'
  retries: 10
  delay: 2
  tags:
    - reconfigure
    - restart

# Failsafe for when HAproxy External docker is timing out externally
# Happens randomly on certain cloud providers
- name: Check if HAProxy External is timing out
  block:
    - name: Check if HAProxy External is timing out
      uri:
        url: "{{ external_url_sanitised }}"
        timeout: 5
      register: haproxy_ext_res
      failed_when: haproxy_ext_res.status == -1
  rescue:
    - name: Restart HAProxy External docker on failure
      docker_container:
        name: haproxy
        restart: true
  when: "'haproxy_external' in group_names"
  tags:
    - reconfigure
    - restart
